source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -1358,85 +1358,66 @@ It will be refreshed by the next periodic check or,
-
-    def is_notmaster(message):
-        if is_recovering(message):
-            return false
-        return ("not master" in message)
-
-    def is_notmaster_or_recovering(message):
-        return is_recovering(message) or is_notmaster(message)
-
-    def parse_gle(response):
-        if "err" in response:
-            if is_notmaster_or_recovering(response["err"]):
-                handle_not_master_or_recovering(response["err"])
-
-    # Parse response to any command besides getLastError.
-    def parse_command_response(response):
-        if not response["ok"]:
-            if is_notmaster_or_recovering(response["errmsg"]):
-                handle_not_master_or_recovering(response["errmsg"])
-
-    def parse_query_response(response):
-        if the "QueryFailure" bit is set in response flags:
-            if is_notmaster_or_recovering(response["$err"]):
-                handle_not_master_or_recovering(response["$err"])
-
-    def handle_not_master_or_recovering(message):
-        replace server's description with
-        new ServerDescription(type=Unknown, error=message)
-
-        if multi-threaded:
-            request immediate check
-        else:
-            # Check right now if this is "not master", since it might be a
-            # useful secondary. If it's "node is recovering" leave it for the
-            # next full scan.
-            if is_notmaster(message):
-                check failing server
-
-        clear connection pool for server
-
-See the test scenario called
-"parsing 'not master' and 'node is recovering' errors"
-for example response documents.
-
-When the client sees a "not master" or "node is recovering" error
-it MUST replace the server's description
-with a default ServerDescription of type Unknown.
-It MUST store useful information in the new ServerDescription's error field,
-including the error message from the server.
-
-Multi-threaded and asynchronous clients MUST `request an immediate check`_
-of the server.
-Unlike in the "network error" scenario above,
-a "not master" or "node is recovering" error means the server is available
-but the client is wrong about its type,
-thus an immediate re-check is likely to provide useful information.
-
-For single-threaded clients, in the case of a "not master" error, the client
-MUST check the server immediately (see `checking a single server`_).  For a
-"node is recovering" error, single-threaded clients MUST NOT check the server,
+Server errors are detected from a getLastError response, write command
+response, or query response. The following errors indicate a server
+role change:
+
+.. list-table::
+  :header-rows: 1
+
+  * - Error Name
+    - Error Code
+  * - InterruptedAtShutdown
+    - 11600
+  * - InterruptedDueToReplStateChange
+    - 11602
+  * - NotMaster
+    - 10107
+  * - NotMasterNoSlaveOk
+    - 13435
+  * - NotMasterOrSecondary
+    - 13436
+  * - PrimarySteppedDown
+    - 189
+  * - ShutdownInProgress
+    - 91
```

Comment:
How come we don't include the following errors here? 

Error Name | Error Code
-- | --
HostNotFound | 7
HostUnreachable | 6
NetworkTimeout | 89
SocketException | 9001

Reply:
IIUC these are returned by mongos when mongos can't reach a shard - but SDAM doesn't care. SDAM has marked this node as a mongos and it shouldn't set it to Unknown if it sees one of these errors.

Reply:
I added a section to the rationale explaining why we don't use those error codes, and updated "Use error messages to detect "not master" and "node is recovering"".